### PR TITLE
Add Some Missing APIs

### PR DESCRIPTION
**Pull request by TheBrokenRail**
- ```[NSAlert beginSheetModalForWindow:sheetWindow completionHandler:handler]```
- ```[NSSavePanel setDirectoryURL:url]```
- ```[NSSavePanel directoryURL]```
- ```[X11Window showWindowForAppActivation:frame]```
- ```[X11Window hideWindowForAppActivation:frame]```

Note, the ```X11Window``` APIs don't do anything special and just call their ```withActivation``` equivalents, but its better than before.

### CODE REVIEW

File: AppKit/NSSavePanel.subproj/NSSavePanel.m [hunk 1]
```diff
@@ -143,6 +143,14 @@ -(void)setDirectory:(NSString *)directory {
    _directory=directory;
 }
 
+- (void) setDirectoryURL: (NSURL *) url {
+  [self setDirectory: [url path]];
+}
+
+- (NSURL *) directoryURL {
+  return [NSURL fileURLWithPath: [self directory]];
```
Comment (bugaevc) [inline AppKit/NSSavePanel.subproj/NSSavePanel.m:151]: You can just access `_directory` in both cases.

File: AppKit/include/AppKit/NSAlert.h [hunk 2]
```diff
@@ -72,6 +72,7 @@ enum {
 - (void)layout;
 
 - (void)beginSheetModalForWindow:(NSWindow *)window modalDelegate:delegate didEndSelector:(SEL)selector contextInfo:(void *)info;
+- (void) beginSheetModalForWindow: (NSWindow *) sheetWindow completionHandler: (void (^)(NSModalResponse returnCode)) handler
```
Comment (bugaevc) [inline AppKit/include/AppKit/NSAlert.h:75]: This is not even valid syntax, you need a semicolon at the end.

File: AppKit/NSAlert.m [hunk 3]
```diff
@@ -450,6 +450,38 @@ -(void)beginSheetModalForWindow:(NSWindow *)window modalDelegate:delegate didEnd
...
+}
+
+@end
+
+- (void) beginSheetModalForWindow: (NSWindow *) sheetWindow completionHandler: (void (^)(NSModalResponse returnCode)) handler {
+  [_window setStyleMask:NSDocModalWindowMask];
+  [self layoutIfNeeded];
+  [_window setDefaultButtonCell:[[_buttons objectAtIndex:0] cell]];
+
+  [NSApp beginSheet: _window modalForWindow: sheetWindow modalDelegate: [[_NSAlertCompletionHandler alloc] init: handler] didEndSelector: @selector(sheetDidEnd:returnCode:contextInfo:) contextInfo: 0];
```
Comment (bugaevc) [inline AppKit/NSAlert.m:482]: Here's a problem: nothing releases this newly allocated object.

File: AppKit/include/AppKit/NSAlert.h (hunk 2 above)
Comment (TheBrokenRail) [reply AppKit/include/AppKit/NSAlert.h]: Fixed!

File: AppKit/NSAlert.m (hunk 3 above)
Comment (TheBrokenRail) [reply AppKit/NSAlert.m:482]: How do I do this?

Comment (TheBrokenRail) [reply AppKit/NSAlert.m:482]: ```[self release]``` in ```alertDidEnd```?

File: AppKit/NSSavePanel.subproj/NSSavePanel.m (hunk 1 above)
Comment (TheBrokenRail) [reply AppKit/NSSavePanel.subproj/NSSavePanel.m:151]: Why not just use the functions though?

File: AppKit/NSAlert.m (hunk 3 above)
Comment (bugaevc) [reply AppKit/NSAlert.m:482]: Yes, something like that.

File: AppKit/NSSavePanel.subproj/NSSavePanel.m (hunk 1 above)
Comment (bugaevc) [reply AppKit/NSSavePanel.subproj/NSSavePanel.m:151]: No, it's fine this way too.

File: AppKit/NSAlert.m (hunk 3 above)
Comment (TheBrokenRail) [reply AppKit/NSAlert.m:483]: Fixed!

**Comment by TheBrokenRail** [comment]
Any updates?

**Comment by TheBrokenRail** [comment]
Rebased!

**Comment by bugaevc** [comment]
Squash, please

**Comment by TheBrokenRail** [comment]
Done!

File: AppKit/X11.backend/X11Window.m [hunk 4]
```diff
@@ -474,11 +474,11 @@ - (void) setLevel: (int) value {
 }
 
 - (void) showWindowForAppActivation: (O2Rect) frame {
-    NSUnimplementedMethod();
+    [self showWindowWithoutActivation];
 }
 
 - (void) hideWindowForAppDeactivation: (O2Rect) frame {
-    NSUnimplementedMethod();
+    [self hideWindow];
```
Comment (bugaevc) [inline AppKit/X11.backend/X11Window.m:481]: I don't actually know if we should respect these (OS X style) window behaviors; or if we should just mark utility windows as such and let the window manager deal with it (which we try to do already).

Opinions?

File: AppKit/NSAlert.m [hunk 5]
```diff
@@ -518,23 +518,57 @@ - (void) sheetDidEnd: (NSWindow *) sheet
...
 
-- (void) beginSheetModalForWindow: (NSWindow *) window
-                    modalDelegate: delegate
-                   didEndSelector: (SEL) selector
-                      contextInfo: (void *) info
+@interface _NSAlertCompletionHandler : NSObject {
+    (void (^)(NSModalResponse returnCode)) completionHandler;
+}
+
+- (id) init: (void (^)(NSModalResponse returnCode)) handler;
```
Comment (bugaevc) [inline AppKit/NSAlert.m:525]: The selector should be `initWithBlock:`

File: AppKit/NSAlert.m [hunk 6]
```diff
@@ -518,23 +518,57 @@ - (void) sheetDidEnd: (NSWindow *) sheet
...
+          returnCode: (NSInteger) returnCode
+         contextInfo: (void *) contextInfo;
+
+@end
+
+@implementation _NSAlertCompletionHandler
+
+- (id) init: (void (^)(NSModalResponse returnCode)) handler {
+    self = [super init];
+    completionHandler = handler;
```
Comment (bugaevc) [inline AppKit/NSAlert.m:537]:
```suggestion
    completionHandler = [handler copy];
```

(and release it below)

Comment (TheBrokenRail) [reply AppKit/NSAlert.m]: Done!

File: AppKit/NSAlert.m (hunk 5 above)
Comment (TheBrokenRail) [reply AppKit/NSAlert.m]: Done!

File: AppKit/X11.backend/X11Window.m (hunk 4 above)
Comment (TheBrokenRail) [reply AppKit/X11.backend/X11Window.m:481]: Honestly, I have no idea whats these methods are supposed to do, so I just copied the Win32 implementation, I means if it trying to show the window and do something else, at least it should show the window.

Comment (bugaevc) [reply AppKit/X11.backend/X11Window.m:481]: Right, showing the window should probably just show it. Hiding, not so sure.

File: AppKit/NSAlert.m [hunk 7]
```diff
@@ -518,23 +518,62 @@ - (void) sheetDidEnd: (NSWindow *) sheet
     [self release];
 }
 
-- (void) beginSheetModalForWindow: (NSWindow *) window
-                    modalDelegate: delegate
-                   didEndSelector: (SEL) selector
-                      contextInfo: (void *) info
+@interface _NSAlertCompletionHandler : NSObject {
+    (void (^)(NSModalResponse returnCode)) completionHandler;
```
Comment (bugaevc) [inline AppKit/NSAlert.m:522]: Does this compile? The proper syntax is:

```suggestion
    void (^completionHandler)(NSModalResponse returnCode);
```

See [fuckingblocksyntax.com](http://fuckingblocksyntax.com/) 😆

File: AppKit/X11.backend/X11Window.m (hunk 4 above)
Comment (TheBrokenRail) [reply AppKit/X11.backend/X11Window.m:481]: Why not? What exactly are these method supposed to do?

Comment (bugaevc) [reply AppKit/X11.backend/X11Window.m:481]: Hide the window for app deactivation, just as it says. That's what happens on OS X when you switch to another application: panels get hidden. I don't know if it's done by the app or by the WindowServer on OS X; but in Cocotron the same functionality is to be implemented in the AppKit backend. But again, I'm not sure if we want to reproduce that OS X behavior, or rather stick with whatever the Linux desktop does. And I think I'm liking the latter option more.

Comment (TheBrokenRail) [reply AppKit/X11.backend/X11Window.m]: Done!

File: AppKit/NSAlert.m [hunk 8]
```diff
@@ -518,23 +518,62 @@ - (void) sheetDidEnd: (NSWindow *) sheet
     [self release];
 }
 
-- (void) beginSheetModalForWindow: (NSWindow *) window
-                    modalDelegate: delegate
-                   didEndSelector: (SEL) selector
-                      contextInfo: (void *) info
```
Comment (bugaevc) [inline AppKit/NSAlert.m:524]: What happened to that method's implementation?

Comment (TheBrokenRail) [reply AppKit/NSAlert.m:524]: How did I do this? Fixing it now.

Comment (TheBrokenRail) [reply AppKit/NSAlert.m:524]: Must have been a rebase.

Comment (TheBrokenRail) [reply AppKit/NSAlert.m]: Fixed!

**Comment by TheBrokenRail** [comment]
Any updates?

**Comment by bugaevc** [comment]
If you pinky-swear you've verified it compiles, I'll push the green button.

**Comment by TheBrokenRail** [comment]
The build failed because of CoreGraphics?
```
In file included from /home/connor/Documents/darling/src/external/Heimdal/lib/krb5/stat-kerberos.m:39:
In file included from /home/connor/Documents/darling/src/external/foundation/include/Foundation/Foundation.h:121:
In file included from /home/connor/Documents/darling/src/external/foundation/include/Foundation/NSGeometry.h:5:
In file included from /home/connor/Documents/darling/src/external/cocotron/CoreGraphics/include/CoreGraphics/CoreGraphics.h:31:
/home/connor/Documents/darling/src/external/cocotron/CoreGraphics/include/CoreGraphics/CGEvent.h:29:57: error: unknown type name 'CGEventRef'; did you mean 'CGFontRef'?
extern void CGEventPost(CGEventTapLocation tapLocation, CGEventRef event);
                                                        ^~~~~~~~~~
                                                        CGFontRef
/home/connor/Documents/darling/src/external/cocotron/CoreGraphics/include/CoreGraphics/CGFont.h:25:44: note: 'CGFontRef' declared here
typedef struct CF_BRIDGED_TYPE(id) O2Font *CGFontRef;
                                           ^
In file included from /home/connor/Documents/darling/src/external/Heimdal/lib/krb5/stat-kerberos.m:39:
In file included from /home/connor/Documents/darling/src/external/foundation/include/Foundation/Foundation.h:121:
In file included from /home/connor/Documents/darling/src/external/foundation/include/Foundation/NSGeometry.h:5:
In file included from /home/connor/Documents/darling/src/external/cocotron/CoreGraphics/include/CoreGraphics/CoreGraphics.h:31:
/home/connor/Documents/darling/src/external/cocotron/CoreGraphics/include/CoreGraphics/CGEvent.h:32:26: error: unknown type name 'CGButtonCount'
                         CGButtonCount buttonCount, boolean_t mouseButtonDown,
                         ^
```

**Comment by TheBrokenRail** [comment]
Just merged in the latest changes (locally), lets see if it builds!

**Comment by TheBrokenRail** [comment]
It builds (after a few changes involving ```NSModalResponse```)!

File: AppKit/NSAlert.m [hunk 9]
```diff
@@ -31,8 +30,44 @@ this software and associated documentation files (the "Software"), to deal in
 #import <AppKit/NSStringDrawer.h>
 #import <AppKit/NSTextField.h>
 #import <AppKit/NSWindow-Private.h>
+#import <AppKit/NSApplication.h>
```
Comment (bugaevc) [inline AppKit/NSAlert.m:33]: Why?

File: AppKit/include/AppKit/NSAlert.h [hunk 10]
```diff
@@ -33,6 +33,8 @@ enum {
     NSAlertThirdButtonReturn = 1002,
 };
 
+typedef NSInteger NSModalResponse;
```
Comment (bugaevc) [inline AppKit/include/AppKit/NSAlert.h:36]: I think this is already typedefed in another header; you should just include that.

Comment (bugaevc) [reply AppKit/include/AppKit/NSAlert.h:36]: Ah, I see you've moved it from there.

File: AppKit/NSAlert.m (hunk 9 above)
Comment (TheBrokenRail) [reply AppKit/NSAlert.m:33]: I think I accidentally moved that when trying to get ```NSModalResponse```.

**Comment by TheBrokenRail** [comment]
It builds and I think the implementations are correct, any other issues?

**Comment by bugaevc** [comment]
There are a couple of remaining issues, but let's go ahead and merge this anyway.

**Pull request merged**